docs/proposals/20230123-azure-service-operator.md

```diff
@@ -0,0 +1,235 @@
+- a checkpoint when installing CAPZ in the framework that ASO is also installed and healthy
+- upgrade tests from a version of CAPZ that does not use ASO to one that does,
+  verifying that CAPZ-managed and unmanaged Azure resources become correctly
+  reflected in ASO after the upgrade
+- a test to verify when a CAPZ upgrade includes an upgrade of the ASO control plane
+- a test to verify when a CAPZ upgrade includes a newer API version of an ASO resource
+
+To catch performance and scalability regressions, one-off tests against CAPZ instances managing many and/or large workload clusters will also be performed. Telemetry will be inspected to ensure the time it takes to reconcile a high volume of clusters and the overhead required in the management cluster to run ASO are still acceptable.
+
+Finally, unit tests will be added to cover tightly-scoped functional changes related to ASO adoption, like error handling.
+
+### Graduation Criteria
+
+ASO integration will not be kept behind a feature flag or matriculate through the usual alpha, beta, and stable phases. Instead, the transition will be made gradually so as to distribute potential impact over time.
+
+Each Azure service package, such as `virtualmachines` or `loadbalancers` can be transitioned separately from using the SDK to ASO. The transition for a package will be made in one atomic change, so the SDK- and ASO-based implementations will never exist side-by-side for a given service. However, there will be a period of time where the SDK and ASO are driving mutually exclusive sets of services at the same time, and that time span may include several releases of CAPZ. If significant regressions are found in a particular service implementation, that service (and others that rely on it) be rolled back in isolation to the SDK-based implementation independently of unaffected services.
+
+Since the transition for each service is expected to take some time, it's worth considering which services to transition first. ASO's resources include some references to Azure resources as references to ASO resources which will impose some necessary ordering. For example, a VirtualNetworksSubnet's `spec.owner` refers to the Kubernetes name of the owning ASO VirtualNetwork, so the `virtualnetworks` service would be implemented to use ASO before the same is done for `subnets`. When no ordering is otherwise necessary, the priority to implement each service will be based on new features which are blocked on newer Azure API versions available in ASO but not the SDK and also community involvement.
```

Comment:
> Since the transition for each service is expected to take some time

Do we want to call out somewhere how this would impact the handling of patch releases for critical fixes or security during the transition period? This may require separate trees once a component is migrated to ASO. I'm not suggesting a premature optimization for it, more of a base (maybe hand-wavey) recognition of how it would be handled.

Reply:
Based on the experimentation I've done so far, I think any changes like that would mostly apply to one version or the other (ASO or non-ASO) since the implementations don't overlap much. If an equivalent fix is necessary in both, then there will be extra work to implement the fix twice, but I'm not sure that will generally be so much that we need to take special measures during this transition period. This might fit a bit better in the "Risks and Mitigations" section, so I'll add something about this there.